query-builder/src/main/java/com/datastax/oss/driver/api/querybuilder/QueryBuilderDsl.java

```diff
@@ -0,0 +1,742 @@
+import com.datastax.oss.driver.internal.querybuilder.select.BinaryArithmeticSelector;
+import com.datastax.oss.driver.internal.querybuilder.select.CastSelector;
+import com.datastax.oss.driver.internal.querybuilder.select.ColumnSelector;
+import com.datastax.oss.driver.internal.querybuilder.select.CountAllSelector;
+import com.datastax.oss.driver.internal.querybuilder.select.DefaultBindMarker;
+import com.datastax.oss.driver.internal.querybuilder.select.DefaultSelect;
+import com.datastax.oss.driver.internal.querybuilder.select.ElementSelector;
+import com.datastax.oss.driver.internal.querybuilder.select.FieldSelector;
+import com.datastax.oss.driver.internal.querybuilder.select.FunctionSelector;
+import com.datastax.oss.driver.internal.querybuilder.select.ListSelector;
+import com.datastax.oss.driver.internal.querybuilder.select.MapSelector;
+import com.datastax.oss.driver.internal.querybuilder.select.OppositeSelector;
+import com.datastax.oss.driver.internal.querybuilder.select.RangeSelector;
+import com.datastax.oss.driver.internal.querybuilder.select.SetSelector;
+import com.datastax.oss.driver.internal.querybuilder.select.TupleSelector;
+import com.datastax.oss.driver.internal.querybuilder.term.BinaryArithmeticTerm;
+import com.datastax.oss.driver.internal.querybuilder.term.CastTerm;
+import com.datastax.oss.driver.internal.querybuilder.term.FunctionTerm;
+import com.datastax.oss.driver.internal.querybuilder.term.OppositeTerm;
+import com.datastax.oss.driver.internal.querybuilder.term.TupleTerm;
+import com.google.common.collect.Iterables;
+import java.util.Arrays;
+import java.util.Map;
+
+/** A Domain-Specific Language to build CQL queries using Java code. */
+public interface QueryBuilderDsl {
+
+  /** The identifier of the built-in {@code writetime} function. */
+  CqlIdentifier WRITETIME = CqlIdentifier.fromCql("writetime");
+
+  /** The identifier of the built-in {@code ttl} function. */
+  CqlIdentifier TTL = CqlIdentifier.fromCql("ttl");
+
+  /** Starts a SELECT query for a qualified table. */
+  static SelectFrom selectFrom(CqlIdentifier keyspace, CqlIdentifier table) {
+    return new DefaultSelect(keyspace, table);
+  }
+
+  /**
+   * Shortcut for {@link #selectFrom(CqlIdentifier, CqlIdentifier)
+   * selectFrom(CqlIdentifier.fromCql(keyspace), CqlIdentifier.fromCql(table))}
+   */
+  static SelectFrom selectFrom(String keyspace, String table) {
+    return selectFrom(CqlIdentifier.fromCql(keyspace), CqlIdentifier.fromCql(table));
+  }
+
+  /** Starts a SELECT query for an unqualified table. */
+  static SelectFrom selectFrom(CqlIdentifier table) {
+    return selectFrom(null, table);
+  }
+
+  /** Shortcut for {@link #selectFrom(CqlIdentifier) selectFrom(CqlIdentifier.fromCql(table))} */
+  static SelectFrom selectFrom(String table) {
+    return selectFrom(CqlIdentifier.fromCql(table));
+  }
+
+  /** Selects all columns, as in {@code SELECT *}. */
+  static Selector getAll() {
```

Comment:
As far as I can tell the `getXXX` methods are only used by other parts of DSL and I don't think we intend them to be used by users directly.  .  i.e. `getAll()` is used by `CanAddSelector.all()`.   I think they kind of pollute the `QueryBuilderDsl` a bit, would it be better to move them to a different interface?

Reply:
They are required for `selectors(...)`, and more importantly for nested use:

```java
selectFrom("foo")
    .product(
        getOpposite(getColumn("a")),
        getSum(getColumn("b"), getColumn("c")));
// SELECT -a*(b+c) FROM foo
```

Reply:
Ah ok, that hadn't occurred to me.

I've thought about it a bit and I think the presence of `isXXX` and `getXXX` is giving me heartburn because it makes it less obvious to the user what to use, increases surface area of the API, and the resulting code you will build will be less readable/fluent.

While I know it isn't our goal to match CQL 1:1, but I think the more usable and fluent we can make the API the better, but that would come with some correctness tradeoffs.  I need to look at the API more to understand it to see if I can think of any alternatives.

Reply:
On the other hand, it's also a safety net since the type system guarantees that you can't use a relation where a selector or term is expected, etc.

I made some basic elements like `raw` and `literal` implement multiple interfaces and they can act as different kind of elements, but I think generalizing this pattern to other selectors would be too much.